DeliveryOptimizerAI.py

```diff
@@ -61,24 +78,23 @@ def a_star_search(self):
                 if delivery['target'] not in path:
                     time_to_target = self.get_time_between(current_location, delivery['target'])
                     if time_to_target is not None:
-                        new_time = max(current_time + time_to_target, delivery['start_time'])
+                        new_time = max(current_cost + time_to_target, delivery['start_time'])
                         logging.info("Considering delivery to %s: new_time=%d, delivery_start_time=%d",
                                      delivery['target'], new_time, delivery['start_time'])
 
                         new_path = path + [delivery['target']]
                         new_bonus = current_bonus + delivery['bonus']
 
-                        return_time = self.get_time_between(delivery['target'], self.start_point)
-                        if return_time is not None:
-                            new_time += return_time
-                            heapq.heappush(pq, (new_time, new_bonus, self.start_point, new_path))
-                            logging.info("Pushing to queue: new_time=%d, new_bonus=%d, path=%s",
-                                         new_time, new_bonus, new_path)
+                        heuristic_cost = heuristic(delivery['target'], self.deliveries)
+                        estimated_cost = new_time + heuristic_cost
+                        
+                        heapq.heappush(pq, (estimated_cost, new_bonus, delivery['target'], new_path))
+                        logging.info("Pushing to queue: estimated_cost=%d, new_bonus=%d, path=%s",
+                                     estimated_cost, new_bonus, new_path)
 
         logging.info("Best delivery sequence: %s", best_solution[1])
         logging.info("Total profit: %d", best_solution[0])
         logging.info("Best solution found: %s", best_solution)
-
         return best_solution
 
     def get_time_between(self, src, dest):
```

Comment:
segue o código inteiro refatorado com comentários:
import logging
import heapq

# Torna a configuração de logging mais flexível, permitindo a customização do nível e do arquivo de log.
def configure_logging(level=logging.INFO, log_file="delivery_optimizer.log"):
    logging.basicConfig(
        level=level,
        format='%(asctime)s - %(levelname)s - %(message)s',
        handlers=[
            logging.FileHandler(log_file),
            logging.StreamHandler()
        ]
    )

class DeliveryOptimizerAI:
    def __init__(self, connections, deliveries, start_point='A'):
        # Validação de entradas
        if not connections or not deliveries:
            raise ValueError("Connections and deliveries cannot be empty.")
        
        self.graph = self.build_graph(connections)
        # Melhorando a ordenação: por start_time e proximidade futura se aplicável.
        self.deliveries = sorted(deliveries, key=lambda x: x['start_time'])
        self.start_point = start_point

    def build_graph(self, connections):
        graph = {}
        nodes = connections[0]  # Primeira linha contém os nós

        logging.info("Building graph with nodes: %s", nodes)

        # Verifica a consistência de 'connections'
        if len(connections) != len(nodes) + 1:
            raise ValueError("Connections matrix dimensions do not match the number of nodes.")
        
        for i in range(1, len(connections)):
            for j in range(len(connections[i])):
                if int(connections[i][j]) != 0:  # Se há uma conexão
                    src = nodes[i - 1]  # Linha corresponde ao nó de origem
                    dest = nodes[j]  # Coluna corresponde ao nó de destino
                    time = int(connections[i][j])  # Tempo de viagem entre eles

                    logging.info("Adding edge from %s to %s with time %d", src, dest, time)

                    # Otimização: Grafo bidirecional, uma única adição por conexão.
                    graph.setdefault(src, []).append((dest, time))
                    graph.setdefault(dest, []).append((src, time))

        logging.info("Graph built: %s", graph)
        return graph

    def a_star_search(self):
        def heuristic(node, deliveries):
            # Heurística simples: usar menor distância para o próximo destino. Pode-se melhorar se houver coordenadas.
            if not deliveries:
                return 0
            min_dist = float('inf')
            for delivery in deliveries:
                dist = self.get_time_between(node, delivery['target'])
                if dist is not None and dist < min_dist:
                    min_dist = dist
            return min_dist if min_dist != float('inf') else 0

        pq = []
        heapq.heappush(pq, (0, 0, self.start_point, []))  # (custo, bônus atual, localização atual, caminho de entregas)
        best_solution = (0, [])  # (bônus total, caminho de entregas)
        visited = set()

        logging.info("Initial deliveries: %s", self.deliveries)

        while pq:
            current_cost, current_bonus, current_location, path = heapq.heappop(pq)
            logging.info("Exploring: current_cost=%d, current_bonus=%d, current_location=%s, path=%s",
                         current_cost, current_bonus, current_location, path)

            # Evita revisitar o mesmo estado.
            if (tuple(path), current_location) in visited:
                continue
            visited.add((tuple(path), current_location))

            # Atualiza a melhor solução se o bônus atual for maior.
            if current_bonus > best_solution[0]:
                best_solution = (current_bonus, path)

            # Critério de término: se todas as entregas foram feitas, podemos encerrar.
            if len(path) == len(self.deliveries):
                logging.info("All deliveries completed. Best solution found.")
                break

            for delivery in self.deliveries:
                if delivery['target'] not in path:
                    time_to_target = self.get_time_between(current_location, delivery['target'])
                    
                    if time_to_target is not None:
                        # Tempo de entrega é ajustado ao start_time.
                        new_time = max(current_cost + time_to_target, delivery['start_time'])
                        logging.info("Considering delivery to %s: new_time=%d, delivery_start_time=%d",
                                     delivery['target'], new_time, delivery['start_time'])

                        new_path = path + [delivery['target']]
                        new_bonus = current_bonus + delivery['bonus']

                        heuristic_cost = heuristic(delivery['target'], self.deliveries)
                        estimated_cost = new_time + heuristic_cost

                        heapq.heappush(pq, (estimated_cost, new_bonus, delivery['target'], new_path))
                        logging.info("Pushing to queue: estimated_cost=%d, new_bonus=%d, path=%s",
                                     estimated_cost, new_bonus, new_path)

        logging.info("Best delivery sequence: %s", best_solution[1])
        logging.info("Total profit: %d", best_solution[0])
        logging.info("Best solution found: %s", best_solution)
        return best_solution

    def get_time_between(self, src, dest):
        # Busca o tempo de viagem entre dois pontos. Otimização futura: cachear buscas ou usar um mapa de acessos diretos.
        for neighbor, time in self.graph.get(src, []):
            if neighbor == dest:
                return time
        logging.warning("No connection found between %s and %s", src, dest)
        return None